### PR TITLE
Fixed deprecation warnings and bugs on page operations

### PR DESCRIPTION
**Pull request by cabidop**
Hi, @angusmcleod! Not sure if you are paying any attention at all to the plugin, but I recently started working with Discourse and found that Landing Pages filled exactly the gap I just needed for a project. Seeing that it was unmaintained I started my own simplified version of your plugin, but this week I decided to give it a go and update yours so here we are now 😄 .

Although I am already using my fork, I would appreciate if you could take a look to this PR to include the changes in the main repository so everyone can benefit from them. The list is not huge as I mainly focused on removing deprecation warnings and bringing the basic functionality back to life:

- Cleaned Ruby and Javascript deprecated code when loading the plugin.
- Fixed the "export page" functionality which solves #4, as some Javascript bits were missing for it to properly trigger.
- Fixed the "import page" feature too. This required a little bit more effort as the modal [needed to be migrated](https://meta.discourse.org/t/converting-modals-from-legacy-controllers-to-new-dmodal-component-api/268057) and there were some mismatches in the Ruby code between the exported and the imported information.
- Small fix to the "delete page" feature to remove the recently deleted page from the list.

There are still some small issues with the save functionality (e.g. you can never remove categories and group once you added them) and I have not reviewed yet the global and remote functionality, but at least the plugin is now on a non-broken state.

If you think it is too much to review in one go let me know as I can definitely make multiple PRs, each commit is more or less self-contained. Thanks a lot in advance and hope that you are still around for this!

### CODE REVIEW

**Comment by angusmcleod** [comment]
@cabidop Thanks! I'll take a closer look as soon as I can.

**Comment by cabidop** [comment]
Hi, @angusmcleod ! Any chance to take a look?

File: app/controllers/landing_pages/landing.rb
Comment (angusmcleod) [inline app/controllers/landing_pages/landing.rb:5]: @cabidop Inheriting the application controller here means inheriting quite a lot of logic. Could you walk me through why you're doing this? Perhaps this is to make the new "inline" pages you've added? Does it work for non-inline pages though?

File: lib/landing_pages/import_export/zip_exporter.rb
Comment (angusmcleod) [inline lib/landing_pages/import_export/zip_exporter.rb:22]: @cabidop Why is this logic being removed?

File: assets/javascripts/discourse/components/modal/import-pages.js.es6
Comment (angusmcleod) [inline assets/javascripts/discourse/components/modal/import-pages.js.es6:7]: @cabidop This modal throws an error when you try to close it.

File: app/views/layouts/landing.html.erb
Comment (angusmcleod) [inline app/views/layouts/landing.html.erb:28]: @cabidop Whether it's this change or some other changes, page styling of non-inline pages seems to be broken. For example if you import pages from https://github.com/paviliondev/pavilion-landing-pages and apply the theme https://github.com/paviliondev/pavilion-landing-theme you see this

Screenshot 2023-09-19 at 10 21 45

File: assets/javascripts/discourse/models/landing-page.js.es6
Comment (angusmcleod) [inline assets/javascripts/discourse/models/landing-page.js.es6:72]: @cabidop Is there a need to change this path?

File: lib/landing_pages/page.rb
Comment (angusmcleod) [inline lib/landing_pages/page.rb:20]: @cabidop Why is this being changed?

**Review by angusmcleod — CHANGES_REQUESTED** [review]
@cabidop Thanks a lot for this! I really appreciate the effort you've put in here. I've left a few specific comments. 

On the whole, before I can support this is a merge that bring the plugin up to date with Discourse, There are a few changes here that seem to not be strictly about making the plugin work, and it's not clear to me whether they're breaking some existing non-inline functionality.

I definitely don't want to discourage you! But there may need to be some more focus on the other functionality you haven't been directly using, before I can support this as a general "fix". Essentially, when we make that claim by merging this I don't want to be signing up for a fresh round of support from folks who want to use other features that are newly broken, or still broken.

But keen to support you here still. If you'd like more specific feedback than I've given, let me know.

File: app/controllers/landing_pages/landing.rb
Comment (cabidop) [reply app/controllers/landing_pages/landing.rb:5]: Not directly related to the inline pages (although it simplified quite a bit the implementation), but checking the [`ActionController::Base` documentation](https://api.rubyonrails.org/classes/ActionController/Base.html) seems that by default the only controller that inherits it is the `ApplicationController`, which seems to also be the approach for the default and official Discourse plugins I checked out (e.g. [`chat`](https://github.com/discourse/discourse/blob/main/plugins/chat/app/controllers/chat/base_controller.rb), [`poll`](https://github.com/discourse/discourse/blob/main/plugins/poll/app/controllers/polls_controller.rb), [`styleguide`](https://github.com/discourse/discourse/blob/main/plugins/styleguide/app/controllers/styleguide/styleguide_controller.rb), [`calendar`](https://github.com/discourse/discourse-calendar/blob/main/app/controllers/discourse_post_event/discourse_post_event_controller.rb), [`cakeday`](https://github.com/discourse/discourse-cakeday/blob/main/app/controllers/discourse_cakeday/cakeday_controller.rb), etc).

On the other hand, I have not seen anything specifically discouraging to inherit `ActionController::Base` directly so let me know your thoughts on this.

File: lib/landing_pages/import_export/zip_exporter.rb
Comment (cabidop) [reply lib/landing_pages/import_export/zip_exporter.rb:22]: It was not being used: if you check line 17 in the same file the only considered attributes are `name`, `path` and `remote`, so the `theme_id` and the `group_ids` conditions were never checked.

This also meant that these two attributes (plus `parent_id` and `category_id`) where never added to the exported data, so using `LandingPages::Page.pages_attrs` instead of `%w(name path remote)` for the loop makes that possible now.

File: assets/javascripts/discourse/models/landing-page.js.es6
Comment (cabidop) [reply assets/javascripts/discourse/models/landing-page.js.es6:72]: It seems so, the [routes.rb](https://github.com/paviliondev/discourse-landing-pages/blob/main/config/routes.rb#L8) file only defines the `POST :page/upload` path, which also matches the `upload` method in the [`PageController`](https://github.com/paviliondev/discourse-landing-pages/blob/main/app/controllers/landing_pages/admin/page.rb#L55).

File: lib/landing_pages/page.rb
Comment (cabidop) [reply lib/landing_pages/page.rb:20]: As both the `discourse_attrs` and the `pages_attrs` were used exclusively to create the `writable_attrs` field, it looked like a good idea to make them usable in the `zip_exporter.rb` file and avoid hardcoding the attributes there (see my [previous comment](https://github.com/paviliondev/discourse-landing-pages/pull/11#discussion_r1329544057)).

Comment (cabidop) [reply lib/landing_pages/page.rb:20]: As both the `discourse_attrs` and the `pages_attrs` were used exclusively to create the `writable_attrs` field, it looked like a good idea to make them usable in the `zip_exporter.rb` file and avoid hardcoding the attributes there (see my [previous comment](https://github.com/paviliondev/discourse-landing-pages/pull/11#discussion_r1329544057)). In any case, the behaviour for the `writable_attrs` was not changed besides the addition of the new `inline` field.

File: app/views/layouts/landing.html.erb
Comment (cabidop) [reply app/views/layouts/landing.html.erb:28]: Absolutely right, seems that both tags are required as the `:mobile/:desktop` is used for the base styles and the `:mobile_theme/:desktop_theme` is used for the imported themes and the "manual" updates on the default one.

File: assets/javascripts/discourse/components/modal/import-pages.js.es6
Comment (cabidop) [reply assets/javascripts/discourse/components/modal/import-pages.js.es6:7]: Good catch, seems that the "cancel" button has a different behaviour than clicking the "X" and it is trying to load a blank page instead of just closing the modal.

File: app/views/layouts/landing.html.erb
Comment (cabidop) [reply app/views/layouts/landing.html.erb:28]: Also the current code it is actually ignoring the page setting for the theme, as all the conditionals use `theme_id.*` instead of `@page.theme_id.*`.

File: app/controllers/landing_pages/landing.rb
Comment (angusmcleod) [reply app/controllers/landing_pages/landing.rb:5]: Yes, typically you would inherit the ApplicationController because you're serving content within Discourse and you want all the logic included in the ApplicationController (e.g. all it's before_actions). However in this case, sometimes we don't want all of that logic running because a landing page will not be running inside of the main Discourse app. If there wasn't a good reason to make this change then best not to make it.

File: lib/landing_pages/import_export/zip_exporter.rb
Comment (angusmcleod) [reply lib/landing_pages/import_export/zip_exporter.rb:22]: Ok, but if we're now also checking `theme_id` and `group_ids` won't we need the logic concerning their appropriate `value`?

File: assets/javascripts/discourse/models/landing-page.js.es6
Comment (angusmcleod) [reply assets/javascripts/discourse/models/landing-page.js.es6:72]: Ok.

File: lib/landing_pages/page.rb
Comment (angusmcleod) [reply lib/landing_pages/page.rb:20]: Ok 👍

**Comment by angusmcleod** [comment]
@cabidop Thanks for the followup. 

Let me know when you've made those changes. I've added standard plugin CI in here as a sanity check. That will just run the rspec suite in the plugin. We'll need to make sure that's passing before we merge this.

Thanks again!

File: lib/landing_pages/import_export/zip_exporter.rb
Comment (cabidop) [reply lib/landing_pages/import_export/zip_exporter.rb:22]: With the update on the importer the "raw" values for the attributes (int for the `theme_id` and array of ints for the `group_ids`) can be handled directly, without further conversions between types.

**Comment by cabidop** [comment]
Hey @angusmcleod, thanks a lot for the review! I have just pushed some additional commits to address your comments. I fully agree on leaving out the new "inline" functionality (I will add it as a standalone change once the basics are fixed), and I rephrased the PR itself to better reflect its intention. It is now clear that it only impacts the page related functionality, leaving the fixes for both the global and remote functionalities for a different PR.

I also tested the spec locally and added two small fixes for it, but I am unsure if the CI pipeline will run before merging the PR (as the file is obviously not yet in the `main` branch).

File: assets/javascripts/discourse/components/modal/import-pages.js.es6
Comment (angusmcleod) [reply assets/javascripts/discourse/components/modal/import-pages.js.es6:7]: Fixed 👍

File: lib/landing_pages/import_export/zip_exporter.rb
Comment (angusmcleod) [reply lib/landing_pages/import_export/zip_exporter.rb:22]: Ok

**Comment by angusmcleod** [comment]
@cabidop Ok! I think we can address further issues with the plugin separately. You've achieved a baseline of functionality here that we should bank. Thanks again! In terms of follow up:

- Would you like to be added as an author of the plugin in the plugin.rb?
- If you want to raise another PR I'll review it as a priority
- I'm going to move this plugin back to "working" on meta.discourse.org

If you're keen to use this for the long haul I can also make you a maintainer. Let's see how we go in that respect.

**Review by angusmcleod — APPROVED** [review]

**Pull request merged**
